generator/README.md

```diff
@@ -0,0 +1,1186 @@
+
+```SDL<Enum Name><Value>```
+
+ in camel case.
+
+
+**Requests:**
+
+  * SDLxxx.h
+  * SDLxxx.m
+  
+**Responses:**
+
+  * SDLxxxResponse.h
+  * SDLxxxResponse.m
+
+*Note: as an exception from rules this kind has the suffix 'Response'*
+ 
+**Structures:**
+
+  * SDLxxx.h
+  * SDLxxx.m
+
+**Enums:**
+
+  * SDLxxx.h
+  * SDLxxx.m
+
+**Notifications:**
+
+  * SDLxxx.h
+  * SDLxxx.m
+
+Where the **xxx** is the correspondent item name.
+
+
+## The License Header
+
+All files should begin with the license information.
+
+```
+/*
+ * Copyright (c) 2017 - [year], SmartDeviceLink Consortium, Inc.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following
+ * disclaimer in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ * Neither the name of the SmartDeviceLink Consortium Inc. nor the names of
+ * its contributors may be used to endorse or promote products derived
+ * from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+```
+Where `[year]` in the copyright line is the current (?) year.
+*Note: for simplisity sake the header removed from Objective-C code snippets in the following sections of this document*
+
+
+### General rules for Objective-C classes
+1. Default initializer in every class 
+```objc
+    - (instancetype)init {
+        if ((self = [super initWithName:SDLRPCFunctionNameRegisterAppInterface])) {
+        }
+        return self;
+    }
+```
+*Pease note the required double brackets in the if statement*
```

Comment:
@yLeonid please reply to this

Reply:
Example:
0. if (x=y) {} // is not a comparison but an assignment 
we can write it as:
1. if (nil != (x=y)) {}
or
2. if ((x=y)) {}
The issue with the #0 is that depending on compiler settings it will produce either a warning in better case or an error in worst.
The #2 is the most common way to avoid compiler warnings/errors.

Steps to reproduce the warning (error):
1. create a default command line objc project in Xcode.
2. copy-paste the following code:
```
#import <Foundation/Foundation.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        int x = 0;
        int y = 10;

        if (x = y) {
            NSLog(@"non zero");
        } else {
            NSLog(@"zero");
        }
    }
    return 0;
}
```
3. compile the project.
4. observe the following warning:
```
main.m:16:15: warning: using the result of an assignment as a condition without parentheses [-Wparentheses]
        if (x = y) {
            ~~^~~
/Users/leonid/ZProjects/tmp/XXX_test/XXX_test/main.m:16:15: note: place parentheses around the assignment to silence this warning
        if (x = y) {
              ^
            (    )
/Users/leonid/ZProjects/tmp/XXX_test/XXX_test/main.m:16:15: note: use '==' to turn this assignment into an equality comparison
        if (x = y) {
              ^
              ==
1 warning generated.
```
5. open the project settings.
6. go to the section: "Apple Clang - Warning Policies"
7. set "Treat Warnings as Errors" to 'yes'
8. compile the project.
9. observe the result:
```
main.m:16:15: error: using the result of an assignment as a condition without parentheses [-Werror,-Wparentheses]
        if (x = y) {
            ~~^~~
/Users/leonid/ZProjects/tmp/XXX_test/XXX_test/main.m:16:15: note: place parentheses around the assignment to silence this warning
        if (x = y) {
              ^
            (    )
/Users/leonid/ZProjects/tmp/XXX_test/XXX_test/main.m:16:15: note: use '==' to turn this assignment into an equality comparison
        if (x = y) {
              ^
              ==
1 error generated.

```

Conclusion:
Since the sdl_ios is a public available sdk (project) and as soon as it is released it is out of hands. We have no opportunity to control the user's compiler settings and in this case the best experience (IMHO) would be avoid as many warnings not saying of errors as possible especially if it takes next to none effort on our side.

On the other hand it is really not required and anyone can suppress any warnings in the project settings and so can the user.

Thank you for understanding.